### ISSUE TITLE
[Feature Request] Allocator state changes

### ISSUE THREAD

**Issue by mbezuljTT** [tenstorrent/tt-metal#15262]
**Is your feature request related to a problem? Please describe.**

In the compiler, we want to be able to run allocator with mockup device (#14000) that has an optional **graph capture COMPILER_TRACE mode - **  that runs allocator and throws exceptions if we are out of space (PR #14957; Issue #14815). 

@ayerofieiev-tt  has raised good concern that graph capture mode, if one used it with normal operation; could create invalid allocator state. 

Additionally, @ayerofieiev-tt has explained how future TTNN ops might be jit compiling couple of kernel versions, running graph captures on it's own, and decide what is the best op configuration for a given scenario. 

If compiler is doing graph capture with COMPILER_TRACE mode, and underlying TTNN is doing multiple op calls to decide what it is going to do - each would be captured and for example compiler would capture peak L1 usage of the worst candidate op. 

Though we are still not in the position to run this exact scenario, it would be good to think about both scenarios
- COMPILER_TRACE that doesnt mess up device allocator state) and 
- TTNN op that might be compiling/running multiple ops to device what to really dispatch
and think holistically about right solution for this problem 

Sharing pseudo codes for each scenario

Compiler scenario
```
OpenMockupDevice()

Try:
    GraphProcessor::begin_graph_capture()
    RunEntireGraph() // or subgraph
    GraphProcessor::end_graph_capture()
Catch (ExceptionOutOfL1 e):
    failed_op = GetFailedOp(e)
    OverrideOpToDRAMInterleaved(failed_op)
    RecompileGraph()
```

TTNN scenario
```
OpenDevice()

GraphProcessor::begin_graph_capture() // outer graph capture should see only RunOp(selected_option)
For each op in graph:
    options = JITCompileMultipleOptions(op)
    option_data = []
    For each option in options:
        ForkAllocator()
        GraphProcessor::begin_graph_capture()
        option_data.insert(EvaluateOption(option))
        GraphProcessor::end_graph_capture()
        RevertAllocatorState()
    selected_option = DecideBestOption(option_data)
    RunOp(selected_option)
 GraphProcessor::end_graph_capture()
```

**Describe the solution you'd like**

I think MVP solution for the compiler scenario would be to have opening mockup device branch of current allocator state, and just dies at the end. The question is what happens with nested graph captures in this case, it's not our primary path to utilize nested graph captures for this scenario. 

For TTNN scenario, I think, we need same thing. Unless we want to avoid `RunOp(select_option)`, but instead continue out of state that is decided as the best option `DecideBestOption(option_data)`. If so, reverting state might be one option, and continuing of state might be the other option. 

I'm not opinionated on the particular solution, and there might be an angle that I'm missing - especially with multiple nested graph captures. 

**Describe alternatives you've considered**
This was alternative PR #14957, if we think proper solution is going to take months, I would suggest we get something like that PR to get going.

**Comment by staylorTT** [comment]
@prajaramanTT This should be raised as P1 as without this feature we do not have a work around for our interface.

**Comment by abhullar-tt** [comment]
> [@prajaramanTT](https://github.com/prajaramanTT) This should be raised as P1 as without this feature we do not have a work around for our interface.

Hey, we had a design review on the changes needed to support this feature. Re timelines, https://github.com/tenstorrent/tt-metal/issues/14000 is higher priority and likely will get started on next week. The Allocator changes requested here will likely not be started until early Jan

**Comment by staylorTT** [comment]
@abhullar-tt What is the latest on both this and [#14000 ](https://github.com/tenstorrent/tt-metal/issues/14000)

**Comment by pgkeller** [comment]
@staylorTT all work on serialization, deserialization, offline builds, virtual device, allocator changes is on hold indefinitely.  we can sync if you have questions.  thanks

**Comment by skhatriTT** [comment]
@pgkeller Do we have an ETA for this feature?

**Comment by pgkeller** [comment]
No eta, this work is not scheduled

**Comment by zzigler-tt** [comment]
@pgkeller should we get this FR added to the roadmap to provide a little more transparency on when the team can get around to this request?

**Comment by skhatriTT** [comment]
@pgkeller would it be possible to get a rough timeline for when this might be scheduled for?

**Comment by pgkeller** [comment]
Progress on this is contingent on runtime's hiring plan as present staffing can't address it. We currently have 4 offers out to expand our team, the 2nd set of two will be the staffing to address this (assuming accepts). ETA to begin work is still a couple months away

**Comment by skhatriTT** [comment]
@pgkeller checking in for an update

**Comment by abhullar-tt** [comment]
Haven't started this but this planned to be picked up end of this year/early next year

**Comment by skhatriTT** [comment]
We currently don’t have a way to capture or inspect allocator state, so we’re unable to handle fragmentation cases yet. I understand this is planned for next year, but is there any chance we could look into prioritizing it sooner @abhullar-tt @pgkeller? and this depends on https://github.com/tenstorrent/tt-metal/issues/14000

C.C.: @tt-mpantic @odjuricicTT please feel free to add more context if I've missed anything

**Comment by abhullar-tt** [comment]
While we don't have a way of handling fragmentation natively through the allocator, we do have an allocator API that gives statistics based on the current state of memory (`allocator->get_statistics` param is DRAM or L1 to get a particular view). I believe @tapspatel has used this in the past as well

**Comment by tapspatel** [comment]
Currently in tt-mlir runtime, this is how we do it (information below is specific for MLIR folks)

From ttrt perspective
https://github.com/tenstorrent/tt-mlir/blob/9420fe93b4f7a095a54d10d87658be7b9d89eeb2/tools/ttrt/common/callback.py#L242

We save the device and call
```
memory_views = device.get_memory_view()
    dram_memory_view = memory_views[ttrt.runtime.MemoryBufferType.DRAM]
    l1_memory_view = memory_views[ttrt.runtime.MemoryBufferType.L1]
    l1_small_memory_view = memory_views[ttrt.runtime.MemoryBufferType.L1_SMALL]
    trace_memory_view = memory_views[ttrt.runtime.MemoryBufferType.TRACE]
```

get_memory_view calls the pybinded function here: https://github.com/tenstorrent/tt-mlir/blob/9420fe93b4f7a095a54d10d87658be7b9d89eeb2/runtime/python/runtime/runtime.cpp#L46

see `getMemoryView`

which calls at this point in C++ runtime `https://github.com/tenstorrent/tt-mlir/blob/9420fe93b4f7a095a54d10d87658be7b9d89eeb2/runtime/lib/ttnn/runtime.cpp#L671`

From here, we construct the required memory map using metal APIs
```
std::unordered_map<tt::runtime::MemoryBufferType, tt::runtime::MemoryView>
      memoryMap;
  ::ttnn::MeshDevice &meshDevice =
      deviceHandle.as<::ttnn::MeshDevice>(DeviceRuntime::TTNN);

  auto dramMemoryView = ::tt::tt_metal::detail::GetMemoryView(
      &meshDevice, ::ttnn::BufferType::DRAM);
  auto l1MemoryView = ::tt::tt_metal::detail::GetMemoryView(
      &meshDevice, ::ttnn::BufferType::L1);
  auto l1SmallMemoryView = ::tt::tt_metal::detail::GetMemoryView(
      &meshDevice, ::ttnn::BufferType::L1_SMALL);
  auto traceMemoryView = ::tt::tt_metal::detail::GetMemoryView(
      &meshDevice, ::ttnn::BufferType::TRACE);

  memoryMap[tt::runtime::MemoryBufferType::DRAM] =
      createMemoryView(dramMemoryView);
  memoryMap[tt::runtime::MemoryBufferType::L1] = createMemoryView(l1MemoryView);
  memoryMap[tt::runtime::MemoryBufferType::L1_SMALL] =
      createMemoryView(l1SmallMemoryView);
  memoryMap[tt::runtime::MemoryBufferType::TRACE] =
      createMemoryView(traceMemoryView);
```

See `::tt::tt_metal::detail::GetMemoryView`

This gives us most information we need
```
memory_dict = {}
    memory_dict["num_banks"] = memory_view.num_banks
    memory_dict["total_bytes_per_bank"] = memory_view.total_bytes_per_bank
    memory_dict[
        "total_bytes_allocated_per_bank"
    ] = memory_view.total_bytes_allocated_per_bank
    memory_dict["total_bytes_free_per_bank"] = memory_view.total_bytes_free_per_bank
    memory_dict[
        "largest_contiguous_bytes_free_per_bank"
    ] = memory_view.largest_contiguous_bytes_free_per_bank
    memory_dict["block_table"] = memory_view.block_table
```

Which you can later analyze to determine memory fraq etc. Note, you do not have to call this from ttrt, you can call the C++ api in mlir runtime directly at any point and get the memory state in memory without any saving to disk.

**Comment by skhatriTT** [comment]
Thanks @tapspatel for the context but @abhullar-tt this is not related to the virtual device.

C.C: @odjuricicTT please feel free to add more context if I missed something

**Comment by rpavlovicTT** [comment]
Reviving this issue. At the moment, there are couple of ops in TTNN library that depend on allocator state while making decisions. E.g. matmul op creating its program config. 

On the other side, tt-mlir compiler runs various `NO_DISPATCH` graph captures to inspect whether an op works and if it fits in L1 potentially. However, `NO_DISPATCH` mode currently doesn't work with allocator, i.e. allocator's state is not updated and any op that pokes this state thinks that everything is free. This greatly affects tt-mlir compiler as it can't trust information gotten through graph capture calls. 

Additionally, we plan on extending current graph queries to query subgraphs (today we only do it for single op graphs), and this feature would be crucial in such scenarios.

So, feature request is to make `NO_DISPATCH` mode work with allocator (or if this is not desirable, create a new run mode). 

Note https://github.com/tenstorrent/tt-metal/issues/14000 is being worked on and soon we should have virtual device support.

### Questions

@abhullar-tt can your team consider this feature request? 

> @ayerofieiev-tt has raised good concern that graph capture mode, if one used it with normal operation; could create invalid allocator state. Additionally, @ayerofieiev-tt has explained how future TTNN ops might be jit compiling couple of kernel versions, running graph captures on it's own, and decide what is the best op configuration for a given scenario.

@ayerofieiev-tt (a year later) do you still have concerns about this feature? 

cc @tt-mpantic @odjuricicTT

**Comment by skhatriTT** [comment]
@abhullar-tt also do we know when is this scheduled for?

**Comment by abhullar-tt** [comment]
> Reviving this issue. At the moment, there are couple of ops in TTNN library that depend on allocator state while making decisions. E.g. matmul op creating its program config.
> 
> On the other side, tt-mlir compiler runs various `NO_DISPATCH` graph captures to inspect whether an op works and if it fits in L1 potentially. However, `NO_DISPATCH` mode currently doesn't work with allocator, i.e. allocator's state is not updated and any op that pokes this state thinks that everything is free. This greatly affects tt-mlir compiler as it can't trust information gotten through graph capture calls.
> 
> Additionally, we plan on extending current graph queries to query subgraphs (today we only do it for single op graphs), and this feature would be crucial in such scenarios.
> 
> So, feature request is to make `NO_DISPATCH` mode work with allocator (or if this is not desirable, create a new run mode).
> 
> Note [#14000](https://github.com/tenstorrent/tt-metal/issues/14000) is being worked on and soon we should have virtual device support.
> 
> ### Questions
> [@abhullar-tt](https://github.com/abhullar-tt) can your team consider this feature request?
> 
> > [@ayerofieiev-tt](https://github.com/ayerofieiev-tt) has raised good concern that graph capture mode, if one used it with normal operation; could create invalid allocator state. Additionally, [@ayerofieiev-tt](https://github.com/ayerofieiev-tt) has explained how future TTNN ops might be jit compiling couple of kernel versions, running graph captures on it's own, and decide what is the best op configuration for a given scenario.
> 
> [@ayerofieiev-tt](https://github.com/ayerofieiev-tt) (a year later) do you still have concerns about this feature?
> 
> cc [@tt-mpantic](https://github.com/tt-mpantic) [@odjuricicTT](https://github.com/odjuricicTT)

when we originally spoke about this request, the idea was that allocator would allow state tracking / rollback. is that still a requirement or is the ask to ensure allocator state is updated during no dispatch modes?

**Comment by rpavlovicTT** [comment]
> when we originally spoke about this request, the idea was that allocator would allow state tracking / rollback. is that still a requirement or is the ask to ensure allocator state is updated during no dispatch modes?

I am not aware of the rollback requirement as I was not involved in the project at the time. On the other hand, memory allocation tracking for the compiler purposes could be useful when we face fragmentation problem. Although, I wouldn't pursue this right now.

However, at the moment, I think it would most useful (and easier) to enable `NO_DISPATCH` graph capture to run with memory allocator.

**Comment by abhullar-tt** [comment]
> > when we originally spoke about this request, the idea was that allocator would allow state tracking / rollback. is that still a requirement or is the ask to ensure allocator state is updated during no dispatch modes?
> 
> I am not aware of the rollback requirement as I was not involved in the project at the time. On the other hand, memory allocation tracking for the compiler purposes could be useful when we face fragmentation problem. Although, I wouldn't pursue this right now.
> 
> However, at the moment, I think it would most useful (and easier) to enable `NO_DISPATCH` graph capture to run with memory allocator.

Thanks for the clarification. Once we have MockDevice pr merged we can address this

**Comment by abhullar-tt** [comment]
hey @rpavlovicTT, @msudumbrekar-TT made the changes to enable allocator being tracked in `NO_DISPATCH` mode. He and I synced with Artem since graph capture is a TTNN concept who mentioned that in `DISPATCH` mode the allocator state should be captured. With MockDevice merged, would it fit your requirements to run graph capture in `DISPATCH` mode while MockDevice is enabled?

**Comment by rpavlovicTT** [comment]
> With MockDevice merged, would it fit your requirements to run graph capture in DISPATCH mode while MockDevice is enabled?

Yes, that is a good idea, I assume that should not be much slower than `NO_DISPATCH` mode today, right? 

### API changes proposal

On top of that, would it be possible to expose AllocatorState object to outside world? What I mean by that is enhance existing `query_op_constraints` API as

```cpp
struct QueryOpConstraintsConfig {
    std::optional<AllocatorState> initial_state = std::nullopt;
    bool capture_final_state = false;
};

struct ConstraintQueryResponse {
    // ... existing fields ...
    std::optional<AllocatorState> final_allocator_state;
};

// New API 
// Overrides allocator state to the config.initial_state if present
// Runs op on the device and captures final_allocator_state at the end 
auto query_op_constraints_with_state(Op op, MeshDevice* device, const QueryOpConstraintsConfig& config, Args&&... args);

// Some manipulation of AllocatorState
AllocatorState create_from_tensor_specs(const std::vector<TensorSpec>& specs, const Allocator* alloc);
void clear_allocator_state(MeshDevice* device);

// Simulate deallocation of a tensor
AllocatorState remove_allocation_from_state(const AllocatorState& state, const TensorSpec& tensor, const Allocator* alloc);
```

This would enable L1/DRAM placement decisions with lookahead - storing state at decision points, trying L1 path, and backtracking to DRAM if downstream ops fail. As far as I've seen, AllocatorState with extract_state() and override_state() already exists, so hopefully this builds on existing infrastructure.


My questions for you are:
1. Does this look feasible and how much of work would it be?
2. Does this carry some side effects or additional risks?

I assumed we'd have single allocator per mock device, correct me if I'm wrong. But then if mock device is a mesh device, then we'd have many allocators. This could get little bit complex. Maybe we can talk about this over Zoom since I lack some knowledge of the area.

I know this comes as an additional work compared to the previous request, so apologies for that.

**Comment by abhullar-tt** [comment]
> Yes, that is a good idea, I assume that should not be much slower than NO_DISPATCH mode today, right?

With MockDevice stubbing out device operations I don't expect that it would be slower. 

>I assumed we'd have single allocator per mock device, correct me if I'm wrong. But then if mock device is a mesh device, then we'd have many allocators. This could get little bit complex. Maybe we can talk about this over Zoom since I lack some knowledge of the area.

Mode mode follows same rules as running on Silicon, there is one allocator per device 

Regarding the second request, Im not familiar with `query_op_constraints`, (perhaps @ayerofieiev-tt you have more knowledge on this) so I can't speak to feasibility of exposing AllocatorState in there.

**Comment by abhullar-tt** [comment]
@rpavlovicTT if the DISAPTCH mode + MockDevice satisfy the first request, would you be able to close out this issue so we could start a new thread on the second

**Comment by rpavlovicTT** [comment]
@abhullar-tt Yes, I will open a new thread, and I will close this one once I validate our workflows work in DISPATCH + Mock mode.

**Comment by fahmedchanTT** [comment]
@rpavlovicTT Were you able to validate this ? 

cc: @skhatriTT
